api/wasm/cpp/proxy_wasm_impl.h

```diff
@@ -0,0 +1,1229 @@
+    return std::hash<T>()(x.first) + std::hash<U>()(x.second);
+  }
+};
+
+struct Tuple3Hash {
+  template <typename T, typename U, typename V> std::size_t operator()(const std::tuple<T, U, V>& x) const {
+    return std::hash<T>()(std::get<0>(x)) + std::hash<U>()(std::get<1>(x)) + std::hash<V>()(std::get<2>(x));
+  }
+};
+
+using HeaderStringPairs = std::vector<std::pair<std::string, std::string>>;
+
+class GrpcCallHandlerBase {
+public:
+  GrpcCallHandlerBase(Context* context) : context_(context) {}
+  virtual ~GrpcCallHandlerBase() {}
+
+  void cancel();
+
+  virtual void onCreateInitialMetadata() = 0;
+  virtual void onSuccess(std::unique_ptr<WasmData> message) = 0;
+  virtual void onFailure(GrpcStatus status, std::unique_ptr<WasmData> error_message) = 0;
+
+private:
+  friend class Context;
+
+  Context* const context_;
+  uint32_t token_;
+};
+
+template<typename Message>
+class GrpcCallHandler : public GrpcCallHandlerBase {
+public:
+  GrpcCallHandler(Context* context) : GrpcCallHandlerBase(context) {}
+  virtual ~GrpcCallHandler() {}
+
+  virtual void onSuccess(Message&& response) = 0;
+
+private:
+  void onSuccess(std::unique_ptr<WasmData> message) override {
+    onSuccess(message->proto<Message>());
+  }
+};
+
+class GrpcStreamHandlerBase {
+public:
+  GrpcStreamHandlerBase(Context* context) : context_(context) {}
+  virtual ~GrpcStreamHandlerBase() {}
+
+  // NB: with end_of_stream == true, callbacks can still occur: reset() to prevent further callbacks.
+  void send(std::string_view message, bool end_of_stream);
+  void close(); // NB: callbacks can still occur: reset() to prevent further callbacks.
+  void reset();
+
+  virtual void onCreateInitialMetadata() = 0;
+  virtual void onReceiveInitialMetadata() = 0;
+  virtual void onReceiveTrailingMetadata() = 0;
+  virtual void onReceive(std::unique_ptr<WasmData> message) = 0;
+  virtual void onRemoteClose(GrpcStatus status, std::unique_ptr<WasmData> error_message) = 0;
+
+private:
+  friend class Context;
+
+  Context* const context_;
+  uint32_t token_;
+};
+
+template<typename Request, typename Response>
+class GrpcStreamHandler : public GrpcStreamHandlerBase {
+public:
+  GrpcStreamHandler(Context* context) : GrpcStreamHandlerBase(context) {}
+  virtual ~GrpcStreamHandler() {}
+
+  void send(const Request& message, bool end_of_stream) {
+    std::string output;
+    if (!message.SerializeToString(&output)) {
+      return;
+    }
+    GrpcStreamHandlerBase::send(output, end_of_stream);
+  }
+
+  virtual void onReceive(Response&& message) = 0;
+
+private:
+  void onReceive(std::unique_ptr<WasmData> message) override {
+    onReceive(message->proto<Response>());
+  }
+};
+
+// Context for a stream.  The distinguished context id == 0 is used for non-stream calls.
+class Context {
+public:
+  explicit Context(uint32_t id) : id_(id) {}
+  virtual ~Context() {}
+
+  uint32_t id() { return id_; }
+
+  static std::unique_ptr<Context> New(uint32_t id); // For subclassing.
```

Comment:
Based on NullVMPlugin It seems like plugin context should be written in Plugin namespace and inherit this Context class. But this New method is declared as static so it could be defined once in Plugin namespace. Is this expected?

Reply:
Fixed it.  Yes, it can't be static if we are going to have multiple plugins.  I converted this to a free function which is now either global in the WASM code or in the unique namespace of each plugin.